nixfmt-tree: refactor impl to use `treefmt.withConfig`

MattSturgeon opened the pull request:
Follow up to #390147 and #384857, this PR refactors the new `nixfmt-tree` package to make use of the new `treefmt.withConfig` wrapper.

This is technically a breaking change:
1. Overriding `settings` will no longer entirely replace all settings, instead settings are now merged as modules.
2. Adding the package to a nix shell will no longer make runtime inputs (such as `nixfmt`) available on the PATH, only the wrapped `treefmt` will be added to the shell's PATH.

There are also non-breaking changes:
1. Introduced `runtimeInputs`, for consistency with `treefmt.withConfig` and other parts of nixpkgs, such as `writeShellApplication`.
2. Introduced `nixfmtPackage` to allow specifying a different nixfmt package, without having to know to override `nixfmt-rfc-style`.
3. Deprecated `runtimePackages` with a warning. Use the new args instead.

We could have it so that runtime inputs still end up being installed to `$out/bin/`, but this would negate much of the advantage of delegating most impl-details to `treefmt.withConfig`. In hindsight, it also feels like a bit of a leaky abstraction?

Fixes #395045

cc @infinisil @NixOS/nix-formatting

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [x] [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests): `nix-build -A nixfmt-tree.tests`
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

MattSturgeon committed:
nixfmt-tree: refactor impl to use `treefmt.withConfig`

Refactor the new `nixfmt-tree` package, to make use of the new
`treefmt.withConfig` wrapper.

Breaking changes:
1. Overriding `settings` will no longer entirely replace all settings,
   instead settings are now merged as modules.
2. Adding the package to a nix shell will no longer make runtime inputs
   (such as `nixfmt`) available on the PATH, only the wrapped `treefmt`
   will be added to the shell's PATH.

Non-breaking changes:
1. Introduced `runtimeInputs`, for consistency with `treefmt.withConfig`
   and other parts of nixpkgs, such as `writeShellApplication`.
2. Introduced `nixfmtPackage` to allow specifying a different nixfmt
   package, without having to know to override `nixfmt-rfc-style`.
3. Deprecated `runtimePackages` with a warning. Use the new args
   instead.

infinisil commented on pkgs/by-name/ni/nixfmt-tree/package.nix:
```diff
@@ -62,7 +80,7 @@ buildEnv {
 
         You can then also use `treefmt` in a pre-commit/pre-push [Git hook](https://git-scm.com/docs/githooks)
-        and `nixfmt` with your editors format-on-save feature.
+        and with your editor's format-on-save feature.
```
Yeah so this was the original motivation for including the formatter binaries as well, but as explained in the last formatting team meeting, people should be using treefmt even for format-on-save, because treefmt knows about the options you format the codebase with :)

infinisil reviewed (approved):
LGTM!

MattSturgeon commented:
Any objections to getting this merged @NixOS/nix-formatting ?

jfly commented on pkgs/by-name/ni/nixfmt-tree/package.nix:
```diff
@@ -1,45 +1,63 @@
+        _file = ./package.nix;
+
+        # Log level for files treefmt won't format
+        # The default is warn, which would be too annoying for people who just care about Nix
+        on-unmatched = lib.mkOptionDefault "info";
```
Not your fault, but hasn't this changed in recent versions of treefmt? https://github.com/numtide/treefmt/issues/496

MattSturgeon replied:
Looking at the docs, I believe the default is currently "warn".

I assume "info" is also visible at the default log level, but is probably formatted less obnoxiously than warnings?

jfly reviewed (approved):
One thought for you inline. I trust your decision either way.

MattSturgeon commented on pkgs/by-name/ni/nixfmt-tree/package.nix:
```diff
@@ -1,45 +1,63 @@
+      }
+    ];
 
+    runtimeInputs =
+      lib.warnIfNot hasNixfmt
+        "nixfmt-tree: you have supplied a `runtimeInputs` that does not contain a `nixfmt` package. Consider using `treefmt.withConfig` instead of overriding `nixfmt-tree`."
+        runtimeInputs';
```
I think that's reasonable.

I originally went down that route to match the old API, but since we're adding warnings _anyway_, I could do something like:

- If `runtimePackages` is supplied, a deprecation warning is printed and all runtime inputs are replaced.
- If `runtimeInputs` is supplied, they are appended to `[ nixfmt-rfc-style ]`.

I'm unsure on `runtimeInputs` vs `extraRuntimeInputs`, though. Clarity vs verbosity, I guess.

jfly replied:
Actually, here's another idea that I prefer:

- Leave this as `runtimeInputs`
- Remove `nixfmt-rfc-style` from `runtimeInputs`. Instead, set `formatter.nixfmt.command = lib.getExe nixfmt-rfc-style` above.

That means people cannot get out of using nixfmt here (nor can they break it), which feels appropriate given the name. If people want to use treefmt without nixfmt, they can still use `pkgs.treefmt.withConfig`.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 391319`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixfmt-tree</li>
  </ul>
</details>

MattSturgeon commented:
Based on @jfly's [feedback](https://github.com/NixOS/nixpkgs/pull/391319#discussion_r2019248186), I've reworked the `runtimePackages` -> `runtimeInputs` migrations a little:

I introduced a `nixfmtPackage` argument. This allows users to modify or replace the nixfmt package without having to explicitly override `nixfmt-rfc-style`. I'm open to alternative names or approaches here.

The (old) `runtimePackages` argument defaults to `[ nixfmtPackage ]` and the (new) `runtimeInputs` argument defaults to `[ ]`.

The _final_ `runtimeInputs` supplied to `treefmt.withConfig` is `runtimePackages ++ runtimeInputs`. Once `runtimePackages` is removed, that'll just be `[ nixfmtPackage ] ++ runtimeInputs`.

If a user _explicitly_ overrides `runtimePackages`, they will see a deprecation warning:

```
nix-repl> pkgs.nixfmt-tree.override { runtimePackages = []; }
evaluation warning: nixfmt-tree: overriding `runtimePackages` is deprecated, use `runtimeInputs` instead.
                    Note: you do not need to supply a nixfmt package when using `runtimeInputs`, however you can override `nixfmtPackage` to a different nixfmt package.
                    For additional flexibility, or to configure treefmt without nixfmt, consider using `treefmt.withConfig` instead of `nixfmt-tree`.
«derivation /nix/store/l4xmpvzzwrmhxn3bmmicw0wpa2rh4ngh-nixfmt-tree.drv»
```

jfly reviewed (approved):
LGTM! I was surprised to see the binary name `"nixmt"` still hardcoded in this PR, rather than using `lib.getExe`, but IIUC, it has to be this way for backwards compatibility. I believe this can get cleaner in the future when we drop support for backwards compatiblity.

Thanks for iterating on this!

MattSturgeon commented:
> I was surprised to see the binary name "nixmt" still hardcoded in this PR, rather than using `lib.getExe`

Currently we're running `nixfmt` from the PATH; having added it to the PATH via `runtimeInputs` and `makeWrapper`.

Not adding nixfmt to the PATH and instead configuring its absolute path (via `lib.getExe`) is also an option. I think the migration for `runtimePackages` → `runtimeInputs` may be more work, or handle less edge-cases, but I'm open to trying it out if you prefer.

jfly commented:
> but I'm open to trying it out if you prefer.

No, let's merge this. We can revisit this in the future when we drop the deprecated codepath.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/formatting-team-meeting-2025-04-01/62524/1

awwpotato commented:
@MattSturgeon could this be backported to 24.11?
